### PR TITLE
add a general structure for wrapping up DF JSON, refactor dependents and filers to use it, and use it to wrap 1099 INT data

### PR DESCRIPTION
**Pull request by mrotondo**
## Link to pivotal/JIRA issue
- https://codeforamerica.atlassian.net/browse/FYST-950
## Is PM acceptance required?
- No - merge after code review approval
**Reminder**: merge main into this branch and get green tests before merging to main
## What was done?
- I made a DfJsonAccessor class that mirrors the structure we use to wrap the DF XML, and created subclasses of it to access filers, dependents and 1099-INTs that we extract from the DF JSON
## How to test?
- Automated tests should validate that the 1099-INT data is being captured correctly, and that the refactor of the filer & dependent accessors is not changing any functionality

### CODE REVIEW

File: app/models/df_json_accessor.rb
Comment (powersurge360) [inline app/models/df_json_accessor.rb:7]: I don't 100% like the use of the word `accessor` because to me it implies an api parity with `attr_accessor` somewhere. It is internally consistent with our xml accessor though so I'll not block on it but I did want to say that explicitly.

File: app/models/df_json_accessor.rb
Comment (powersurge360) [inline app/models/df_json_accessor.rb:32]: You may consider using 

```ruby
keys = selectors[selector_name][:key].split(' ')
@json.dig(*keys)
```

Which will enable deep traversal of the JSON structure without an API change.

You may or may not need to splat. You'll have to play with it to see. [Documentation linked for convenience](https://apidock.com/ruby/Hash/dig).

File: app/models/state_file_base_intake.rb
Comment (powersurge360) [inline app/models/state_file_base_intake.rb:77]: I'm surprised to see this notation. I expected it to be a hash. How do we get to object? Am I overlooking something?

**Review by powersurge360 — COMMENTED** [review]
Have a question and once answered I'll approve.

File: app/models/df_json_accessor.rb
Comment (mrotondo) [reply app/models/df_json_accessor.rb:32]: awesome - that's exactly what i wanted and just got lazy and didn't look up how to do it/think of using dig :D

File: app/models/state_file_base_intake.rb
Comment (mrotondo) [reply app/models/state_file_base_intake.rb:77]: If you're talking about the `spouse_filer.first_name` notation - that's a side effect of using the DfJsonAccessor (soon to be renamed) wrapper, as the `define_method` calls create a method "API" to the underlying data

File: app/models/df_json_accessor.rb
Comment (powersurge360) [reply app/models/df_json_accessor.rb]: Oh, I mean in the sense that since:

```ruby
attr_accessor :foo, :bar, :baz
attr_reader :foo, :bar, :baz
attr_writer :foo, :bar, :baz
```

I'd expect to be able to

```ruby
json_accessor foo: { type: :string, key: 'some path' }
```

And the equivalent read/write APIs. But it's consistent with how we're doing xml stuff and I don't want to _block_ on it, necessarily.

File: app/models/state_file_base_intake.rb
Comment (powersurge360) [reply app/models/state_file_base_intake.rb:77]: Ah, I see. I was thinking `direct_file_json_data` was the entry point to the json but it's actually an object with many entry points. Cool cool cool.

File: spec/models/df_json_wrapper_spec.rb
Comment (powersurge360) [inline spec/models/df_json_wrapper_spec.rb:19]: NIT: In some editors, if you use the name of the format for the heredoc (eg, `JSON` here), it will engage the LSP and give goodies like syntax highlighting and even linting/completion

image

**Review by powersurge360 — APPROVED** [review]
Seems good. Minor note about heredoc that may not affect anyone but me anyways lol

File: spec/models/df_json_wrapper_spec.rb
Comment (mrotondo) [reply spec/models/df_json_wrapper_spec.rb:19]: oooh, it works in rubymine too! thanks, that's awesome.

File: app/models/df_json_accessor.rb
Comment (mrotondo) [reply app/models/df_json_accessor.rb]: ah, gotcha! yes, that's _much_ cleaner & i've switched over to that syntax. thanks again for the extremely thoughtful review, and for getting into detail about your suggestions (loved the screenshot explaining the heredoc change - i've never used the <<-/<<~ syntax before so it helped me understand what you meant)!

Comment (mrotondo) [reply app/models/df_json_accessor.rb]: and of course, lmk if i was still not quite understanding what you were thinking

Comment (mrotondo) [reply app/models/df_json_accessor.rb]: in particular, i've seen you put multiple attribute names on a attr_reader/writer/accessor line before, but i wonder how you feel about that style now that each `json_reader` specification is a whole line unto itself. i switched from having multiple `json_reader`s to just one, but could switch back if you think it'd be clearer

File: app/models/state_file_base_intake.rb
Comment (mpidcock) [inline app/models/state_file_base_intake.rb:102]: @mrotondo noting here that there's another PR about to be merged that is removing this `humanize` and to look out for the conflict when merging. CC @jenny-heath

Comment (mpidcock) [reply app/models/state_file_base_intake.rb:102]: This could be avoided if you also remove it in this PR

File: app/models/df_json_accessor.rb
Comment (powersurge360) [reply app/models/df_json_accessor.rb]: I may be inclined to break the hashes up over multiple lines but having them all be one call seems alright to me. Yea, this is exactly the sort of thing I was thinking about and looks good to me. Ultimately, either way you're having to set up a hash of keys -> configuration so may as well make declaring it and registering it a single step.

**Review by powersurge360 — APPROVED** [review]
Like the new accessor formatting, looks good.

**Pull request merged**
